pkg/collector/corechecks/network-devices/cisco-sdwan/report/metrics.go

```diff
@@ -0,0 +1,308 @@
+		}
+		ms.sender.Gauge(ciscoSDWANMetricPrefix+"control_connection.status", float64(status), "", tags)
+	}
+}
+
+// SendOMPPeerMetrics send OMP Peer state metrics
+func (ms *SDWanSender) SendOMPPeerMetrics(ompPeers []client.OMPPeer) {
+	for _, entry := range ompPeers {
+		deviceTags := ms.getDeviceTags(entry.VmanageSystemIP)
+		remoteTags := ms.getRemoteDeviceTags(entry.Peer)
+
+		tags := append(deviceTags, remoteTags...)
+		tags = append(tags, "legit:"+entry.Legit, "refresh:"+entry.Refresh, "type:"+entry.Type, "state:"+entry.State)
+
+		status := 0
+		if entry.State == "up" {
+			status = 1
+		}
+		ms.sender.Gauge(ciscoSDWANMetricPrefix+"omp_peer.status", float64(status), "", tags)
+	}
+}
+
+// SendBFDSessionMetrics sends BFD session state metrics
+func (ms *SDWanSender) SendBFDSessionMetrics(bfdSessionsState []client.BFDSession) {
+	for _, entry := range bfdSessionsState {
+		deviceTags := ms.getDeviceTags(entry.VmanageSystemIP)
+		remoteTags := ms.getRemoteDeviceTags(entry.SystemIP)
+
+		tags := append(deviceTags, remoteTags...)
+		tags = append(tags, "local_color:"+entry.LocalColor, "remote_color:"+entry.Color, "proto:"+entry.Proto, "state:"+entry.State)
+
+		status := 0
+		if entry.State == "up" {
+			status = 1
+		}
+		ms.sender.Gauge(ciscoSDWANMetricPrefix+"bfd_session.status", float64(status), "", tags)
+	}
+}
+
+// SendDeviceCountersMetrics sends device counters metrics
+func (ms *SDWanSender) SendDeviceCountersMetrics(deviceCounters []client.DeviceCounters) {
+	for _, entry := range deviceCounters {
+		tags := ms.getDeviceTags(entry.SystemIP)
+
+		ms.sender.MonotonicCount(ciscoSDWANMetricPrefix+"crash.count", float64(entry.CrashCount), "", tags)
+		ms.sender.MonotonicCount(ciscoSDWANMetricPrefix+"reboot.count", float64(entry.RebootCount), "", tags)
+	}
+}
+
+// Commit commits to the sender
+func (ms *SDWanSender) Commit() {
+	ms.sender.Commit()
+	ms.expireTimeSent()
+}
+
+// gaugeWithTimestamp wraps sender GaugeWithTimestamp with error handling
+func (ms *SDWanSender) gaugeWithTimestamp(name string, value float64, tags []string, ts float64) {
+	err := ms.sender.GaugeWithTimestamp(name, value, "", tags, ts)
+	if err != nil {
+		log.Warnf("Error sending Cisco SD-WAN metric %s : %s", name, err)
+	}
+}
+
+// countWithTimestamp wraps sender CountWithTimestamp with error handling
+func (ms *SDWanSender) countWithTimestamp(name string, value float64, tags []string, ts float64) {
+	err := ms.sender.CountWithTimestamp(name, value, "", tags, ts)
+	if err != nil {
+		log.Warnf("Error sending Cisco SD-WAN metric %s : %s", name, err)
+	}
+}
+
+func (ms *SDWanSender) getMetricKey(metric string, tags []string) string {
+	return metric + ":" + strings.Join(tags, ",")
+}
+
+func (ms *SDWanSender) shouldSendEntry(key string, ts float64) bool {
+	lastTs, ok := ms.lastTimeSent[key]
+	if ok && lastTs >= ts {
+		return false
+	}
+	return true
+}
+
+func (ms *SDWanSender) setNewSentTimestamp(newTimestamps map[string]float64, key string, ts float64) {
```

Comment:
Good catch, this shouldn't be  method of `SDWanSender`, moving to a util
We unfortunately cannot modify `ms.lastTimeSent` here because of the following case:
Say the lastSentTimestamp is 10. We receive a set of metrics with timestamps 8, 10, 14 and 12.
8 and 10 are discarded since lower than the last sent timestamp, but if we update `ms.lastTimeSent` while iterating over the new samples, 12 will also get discarded (since they are not in order)